flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,573 @@
+import java.net.URL;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.core.Is.is;
+import static org.hamcrest.core.IsNot.not;
+import static org.junit.Assert.fail;
+
+/** {@code PackagedProgramRetrieverImplTest} tests {@link PackagedProgramRetrieverImpl}. */
+public class PackagedProgramRetrieverImplTest extends TestLogger {
+
+    @Rule public ClasspathProvider classpathProvider = new ClasspathProvider();
+
+    @Test
+    public void testDeriveEntryClassInformationForPythonBasedOnJobName() throws FlinkException {
+        // TODO: see FLINK-23154
+        Assume.assumeTrue(
+                "This tests only succeeds if " + ConfigConstants.ENV_FLINK_OPT_DIR + " is set.",
+                System.getProperty(ConfigConstants.ENV_FLINK_OPT_DIR) != null);
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null,
+                        Collections.emptyList(),
+                        null,
+                        PackagedProgramUtils.getPythonDriverClassName(),
+                        new String[0]);
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(PackagedProgramUtils.getPythonDriverClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJarFile().get(),
+                is(PackagedProgramUtils.getPythonJar().getFile()));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationForPythonBasedOnParameter() throws FlinkException {
+        // TODO: see FLINK-23154
+        Assume.assumeTrue(
+                "This tests only succeeds if " + ConfigConstants.ENV_FLINK_OPT_DIR + " is set.",
+                System.getProperty(ConfigConstants.ENV_FLINK_OPT_DIR) != null);
+
+        final EntryClassInformationProvider informationProvider =
+                PackagedProgramRetrieverImpl.createEntryClassInformationProvider(
+                        null, Collections.emptyList(), null, null, new String[] {"--python"});
+
+        assertThat(informationProvider.getJobClassName().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJobClassName().get(),
+                is(PackagedProgramUtils.getPythonDriverClassName()));
+        assertThat(informationProvider.getJarFile().isPresent(), is(true));
+        assertThat(
+                informationProvider.getJarFile().get(),
+                is(PackagedProgramUtils.getPythonJar().getFile()));
+    }
+
+    @Test
+    public void testDeriveEntryClassInformationForCustomJar()
```

Comment:
What is actually being derived from anything? It seems like all we're testing here is that the stuff passed via the method is eventually assigned.

Reply:
We're testing the code path where a jar is set. Admittedly, it's a quite basic test. But I tried to cover all code paths of `createEntryClassInformationProvider`.